# HG changeset patch
# User Piotr Held <pjheld@gmail.com>
# Date 1509643631 25200
#      Thu Nov 02 10:27:11 2017 -0700
# Branch stable
# Node ID b328ff3ce0f767e6cb2a00d421184d9e2ddfd13c
# Parent  4a46c83bcb7154d842399bb68c4a6b1ff53369dc
call delete method for handle class objects (bug #46497)

* ov-classdef.cc (handle_cdef_object::handle_cdef_object):
Call delete_object to ultimately invoke the delete method for the
object.

* test/bug-46497/bug-46497.tst, test/bug-46497/class_bug46497.m,
test/bug-46497/module.mk: New files.
* test/module.mk: Update.

diff --git a/libinterp/octave-value/ov-classdef.cc b/libinterp/octave-value/ov-classdef.cc
--- a/libinterp/octave-value/ov-classdef.cc
+++ b/libinterp/octave-value/ov-classdef.cc
@@ -1783,16 +1783,19 @@ cdef_object_scalar::is_partially_constru
 inline void
 cdef_object_scalar::mark_as_constructed (const cdef_class& cls)
 {
   ctor_list.erase (cls);
 }
 
 handle_cdef_object::~handle_cdef_object (void)
 {
+  // Call classdef "delete()" method on object
+  get_class ().delete_object (get_class ());
+
 #if DEBUG_TRACE
   std::cerr << "deleting " << get_class ().get_name ()
             << " object (handle)" << std::endl;
 #endif
 }
 
 value_cdef_object::~value_cdef_object (void)
 {
diff --git a/test/bug-46497/bug-46497.tst b/test/bug-46497/bug-46497.tst
new file mode 100644
--- /dev/null
+++ b/test/bug-46497/bug-46497.tst
@@ -0,0 +1,24 @@
+## Copyright (C) 2017 Piotr Held
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 3 of the License, or
+## (at your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <http://www.gnu.org/licenses/>.
+
+%!test
+%! global __bug46497_global__
+%! __bug46497_global__ = 'test_bug46497';
+%! a = class_bug46497 ();
+%! a = [];
+%! assert(__bug46497_global__,'deleted');
\ No newline at end of file
diff --git a/test/bug-46497/class_bug46497.m b/test/bug-46497/class_bug46497.m
new file mode 100644
--- /dev/null
+++ b/test/bug-46497/class_bug46497.m
@@ -0,0 +1,8 @@
+classdef class_bug46497 < handle
+  methods
+    function delete (self)
+      global __bug46497_global__
+      __bug46497_global__ = 'deleted';
+    endfunction
+  endmethods
+endclassdef
diff --git a/test/bug-46497/module.mk b/test/bug-46497/module.mk
new file mode 100644
--- /dev/null
+++ b/test/bug-46497/module.mk
@@ -0,0 +1,5 @@
+bug_46497_TEST_FILES = \
+  %reldir%/bug-46497.tst \
+  %reldir%/class_bug46497.m
+
+TEST_FILES += $(bug_46497_TEST_FILES)
diff --git a/test/module.mk b/test/module.mk
--- a/test/module.mk
+++ b/test/module.mk
@@ -51,16 +51,17 @@ DIRSTAMP_FILES += %reldir%/$(octave_dirs
 
 include %reldir%/bug-35448/module.mk
 include %reldir%/bug-35881/module.mk
 include %reldir%/bug-36025/module.mk
 include %reldir%/bug-38236/module.mk
 include %reldir%/bug-38691/module.mk
 include %reldir%/bug-41723/module.mk
 include %reldir%/bug-44940/module.mk
+include %reldir%/bug-46497/module.mk
 include %reldir%/bug-46660/module.mk
 include %reldir%/bug-49379/module.mk
 include %reldir%/bug-50014/module.mk
 include %reldir%/bug-50035/module.mk
 include %reldir%/bug-50716/module.mk
 include %reldir%/bug-51192/module.mk
 include %reldir%/bug-51532/module.mk
 include %reldir%/bug-51534/module.mk
